android/java/org/chromium/chrome/browser/crypto_wallet/fragments/dapps/AddSwitchChainNetworkFragment.java

```diff
@@ -0,0 +1,253 @@
+            btnAddSwitchNetwork.setText(R.string.approve);
+            SpannableString addChainHeading = Utils.createSpannableString(
+                    getString(R.string.brave_wallet_allow_add_network_description),
+                    getString(R.string.brave_wallet), new ClickableSpan() {
+                        @Override
+                        public void onClick(@NonNull View widget) {
+                            TabUtils.openUrlInNewTab(false, Utils.BRAVE_SUPPORT_URL);
+                            TabUtils.bringChromeTabbedActivityToTheTop(getActivity());
+                        }
+                    }, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
+
+            TextView tvAddChainTitle = view.findViewById(R.id.fragment_add_switch_chain_tv_title);
+            tvAddChainTitle.setText(R.string.brave_wallet_allow_add_network_heading);
+
+            TextView addChainDesc = view.findViewById(R.id.fragment_add_switch_chain_tv_text);
+            addChainDesc.setMovementMethod(LinkMovementMethod.getInstance());
+            addChainDesc.setText(addChainHeading);
+
+        } else if (mPanelType == SWITCH_ETHEREUM_CHAIN) {
+            btnAddSwitchNetwork.setText(R.string.brave_wallet_allow_change_network_button);
+        }
+        btnAddSwitchNetwork.setOnClickListener(v -> {
+            if (mPanelType == ADD_ETHEREUM_CHAIN) {
+                processAddChainRequest(mNetworkInfo, true);
+                processNextAddNetworkRequest();
+            } else if (mPanelType == SWITCH_ETHEREUM_CHAIN) {
+                processSwitchChainRequest(mSwitchChainRequest, true);
+                processNextSwitchChainRequest();
+            }
+        });
+        Button btnAddSwitchCancel =
+                view.findViewById(R.id.fragment_add_switch_eth_chain_btn_cancel);
+        btnAddSwitchCancel.setOnClickListener(v -> {
+            if (mPanelType == ADD_ETHEREUM_CHAIN) {
+                processAddChainRequest(mNetworkInfo, false);
+                processNextAddNetworkRequest();
+            } else if (mPanelType == SWITCH_ETHEREUM_CHAIN) {
+                processSwitchChainRequest(mSwitchChainRequest, false);
+                processNextSwitchChainRequest();
+            }
+        });
+        return view;
+    }
+
+    @Override
+    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+        // finishNativeInitialization may be called before the view creation
+        fetchNetworkInfo();
+    }
+
+    @Override
+    public void finishNativeInitialization() {
+        fetchNetworkInfo();
+    }
+
+    private void fetchNetworkInfo() {
+        // check if actiity has the native service and data is being fetched for the first time
+        if (getActivity() == null || !(getActivity() instanceof BraveWalletBaseActivity)
+                || ((BraveWalletBaseActivity) getActivity()).getJsonRpcService() == null
+                || mNetworkInfo != null)
```

Comment:
maybe better to have a boolean var there saying `initialized` and inside a lock, as `mNetworkInfo` is an async call and it could be that it executes that code twice: first after view creations and second after `finishNativeInitialization`

Reply:
Yes, though that should be handled by the `mNetworkInfo != null`, if we have the data then it will [return](https://github.com/brave/brave-core/pull/13185/files/0de02dbb281b947d26d6bca34c48a5cdf2b9dd5d#diff-789ae456934fc725d72f3ac2ad7d0fcb4f7c756a5684f1b1a10ac2b67768bd2cR152)